sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,686 @@
+ */
+static size_t spi_memcpy(const dif_spi_device_t *spi, fifo_ptrs_t *fifo,
+                         uint16_t fifo_base, uint16_t fifo_len,
+                         uint8_t *byte_buf, size_t buf_len, bool is_recv) {
+  uint16_t bytes_left = fifo_bytes_in_use(*fifo, fifo_len);
+  // When sending, the bytes left are the empty space still available.
+  if (!is_recv) {
+    bytes_left = fifo_len - bytes_left;
+  }
+
+  if (bytes_left > buf_len) {
+    bytes_left = buf_len;
+  }
+  if (bytes_left == 0) {
+    return 0;
+  }
+
+  // For recepit, we advance the read pointer, which indicates how far ahead
+  // we've read so far; for sending, we advance the write pointer, which
+  // indicates how far ahead we've written.
+  fifo_ptr_t *ptr;
+  if (is_recv) {
+    ptr = &fifo->read_ptr;
+  } else {
+    ptr = &fifo->write_ptr;
+  }
+
+  uint16_t total_bytes_available = bytes_left;
+
+  // First, bring the fifo write pointer into word alignment, so we can do
+  // simple full-word I/O rather than partial word I/O; note that the FIFO
+  // SRAM is not guaranteed to support subword writes.
+  if (misalign32(ptr->offset) != 0) {
+    // Misalignment and realignment, corresponding to the number of bytes ahead
+    // of the previous alignment boundary, and the number of bytes to make up
+    // for it.
+    //
+    // For exmaple, 0x3 has misalign of 3 and realign of 1.
+    ptrdiff_t misalign = misalign32(ptr->offset);
+    ptrdiff_t realign = sizeof(uint32_t) - misalign;
+    // Note that we might be doing less I/O than the misalignment requires; we
+    // might be off by a single byte, but not have the full three bytes for full
+    // realignment.
+    if (realign > bytes_left) {
+      realign = bytes_left;
+    }
+
+    // Converts `offset`, which points to a subword boundary, to point to the
+    // start of the current word it points into.
+    ptrdiff_t current_word_offset = ptr->offset - misalign;
+    uint32_t current_word =
+        mmio_region_read32(spi->base_addr, fifo_base + current_word_offset);
+
+    // Act on only to a suffix of `current_word`, corresponding to the necessary
+    // realignment.
+    uint8_t *current_subword = ((uint8_t *)&current_word) + misalign;
+    if (is_recv) {
+      memcpy(byte_buf, current_subword, realign);
+    } else {
+      // When sending, we need to commit the staged write.
+      memcpy(current_subword, byte_buf, realign);
+      mmio_region_write32(spi->base_addr, fifo_base + current_word_offset,
+                          current_word);
+    }
+
+    fifo_ptr_increment(ptr, realign, fifo_len);
+    byte_buf += realign;
+    bytes_left -= realign;
+  }
+
+  // Now, we just do full word I/O until we run out of stuff to act on.
+  while (bytes_left > 0) {
+    // At the end, we may not have a full word to copy, but it's otherwise
+    // the same case as a full word, since we're already word aligned (if
+    // this would be a subword read, it would end the loop anyway).
+    uint16_t bytes_to_copy =
```

Comment:
Every other time you've done a clamp you haven't used a ternary, I don't think I'd start here.

Reply:
Good catch.